oak/server/rust/oak_runtime/src/node/grpc/client.rs

```diff
@@ -63,67 +67,134 @@ impl GrpcClientNode {
+            error!("Couldn't connect to {}: {:?}", self.uri, error);
+            invocation.send_error(rpc::Code::NotFound, "Service connection failed", runtime);
+            OakStatus::ErrInternal
+        })?;
+        grpc_client.ready().await.map_err(|error| {
+            error!("Service was not ready: {}", error);
+            invocation.send_error(rpc::Code::NotFound, "Service not ready", runtime);
+            OakStatus::ErrInternal
+        })?;
+
+        let codec = VecCodec::default();
+        let path = request.method_name.parse().map_err(|error| {
+            error!("Invalid URI {}: {}", request.method_name, error);
+            invocation.send_error(rpc::Code::InvalidArgument, "Invalid URI", runtime);
+            OakStatus::ErrInternal
+        })?;
+
+        // Forward the request to the external gRPC service and wait for the response(s).
+        let request = tonic::Request::new(request.req_msg);
+        let request_stream =
+            request.map(|m| futures_util::stream::once(futures_util::future::ready(m)));
+        let mut rsp_stream = grpc_client
+            .streaming(request_stream, path, codec)
+            .await
+            .map_err(|error| {
+                error!("Request failed: {}", error);
+                invocation.send_error(tonic_code_to_grpc(error.code()), error.message(), runtime);
+                OakStatus::ErrInternal
             })?;
-            debug!("Incoming gRPC request: {:?}", request);
-
-            // Send an unary request to an external gRPC service and wait for the response.
-            let response = self
-                .handler
-                .unary_request(request)
-                .await
-                .unwrap_or_else(|error| {
-                    error!("Couldn't send gRPC request: {:?}", error);
-                    GrpcResponse {
-                        rsp_msg: vec![],
-                        status: Some(from_abi_status(error)),
-                        last: true,
-                    }
-                });
-
-            // Send a response back to the invocation channel.
-            debug!("Sending gRPC response: {:?}", response);
-            invocation
-                .send_response(response, &runtime)
-                .map_err(|error| {
-                    error!("Couldn't send gRPC response to the invocation: {:?}", error);
-                    error
-                })?;
+
+        let body_stream = rsp_stream.get_mut();
+        loop {
+            let message = body_stream.message().await.map_err(|error| {
+                error!("Failed to read response: {}", error);
+                invocation.send_error(rpc::Code::Internal, "Failed to read response", runtime);
+                OakStatus::ErrInternal
+            })?;
+
+            if let Some(message) = message {
+                let encap_rsp = GrpcResponse {
+                    rsp_msg: message,
+                    status: None,
+                    last: false,
```

Comment:
Will Oak streaming methods work if there will not be a `last: true` message?

Reply:
They should treat the channel closure as an indication that the last message in the channel was `last: true`; I believe the [C++ code behaved the same](https://github.com/project-oak/oak/blob/2d79f622b622848fd25aa93a1edce672be3bf8c1/oak/server/grpc_client_node.cc#L125)